### PR TITLE
Replace incorrect usages of "NotImplemented"

### LINKED ISSUE #7266: return NotImplemented
This is minor and not urgent, but in batch normalization: https://github.com/pytorch/pytorch/blob/76d3c30783d3e808b070f8350d9f102bb2396944/torch/nn/modules/batchnorm.py#L48

It seems to be `raise NotImplementedError`, according to the [python doc](https://docs.python.org/3.7/library/constants.html#NotImplemented).

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #7266. Replaces "NotImplemented" (which is supposed to be used for
binary ops) with the correct "NotImplementedError".

cc @soumith

### CODE REVIEW

**Review by ssnl — CHANGES_REQUESTED** [review]
`NotImplementedError` should be raised, not returned

**Comment by zou3519** [comment]
Yikes, thanks for the catch @SsnL!

File: torch/distributions/kl.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def _dispatch_kl(type_p, type_q):
     matches = [(super_p, super_q) for super_p, super_q in _KL_REGISTRY
                if issubclass(type_p, super_p) and issubclass(type_q, super_q)]
     if not matches:
-        return NotImplementedError
+        return NotImplemented
```
Comment (SsnL) [inline torch/distributions/kl.py]: why was this reverted?

File: torch/nn/modules/batchnorm.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ def reset_parameters(self):
             self.bias.data.zero_()
 
     def _check_input_dim(self, input):
-        return NotImplemented
+        raise NotImplementedError
```
Comment (apaszke) [inline torch/nn/modules/batchnorm.py]: It should also be `raise NotImplementedError()`. Otherwise you're raising a class as an error, not an instance

**Review by apaszke — DISMISSED** [review]

**Review by ssnl — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
